src/charm.py

```diff
@@ -214,6 +217,17 @@ def _on_collect_unit_status(self, event: ops.CollectStatusEvent) -> None:
         if not self.model.relations.get("cos-agent"):
             event.add_status(BlockedStatus("Grafana Agent is not related"))
 
+        upstream_snap = get_installed_snap_service(UPSTREAM_SNAP)
+
+        # this is necessary when doing a charm upgrade coming from revision 27
+        if upstream_snap.present:
+            event.add_status(
+                BlockedStatus(
+                    "Please add an empty file as resource. For more info see: "
+                    "https://charmhub.io/openstack-exporter#known-issues"
```

Comment:
It's not a code issue but more of the doc issue outside of this repository but I comment about it here.

First of all, It's great to see the initiative to prepare upgrade notes for this kind of migration. Huge thanks for that.

The link states:
https://charmhub.io/openstack-exporter#known-issues
> Upgrading the charm from revision ≤ 27 to a revision higher than 28 may cause the snap to move a blocked state and require manual intervention. This is intentional.

It's a little bit awkward to have <= 27 and > 28 as the condition. I know the rev 28 was a bit problematic in this context but it won't hit to the stable channel and will be overridden by a new revision soon. So from a documentation perspective, it can be simplified as "when upgrading from <= 27" instead of mentioning to a target revision.

> Please see the related tutorial.

There is no link to the actual tutorial. Would be good to add a hyperlink to https://charmhub.io/openstack-exporter/docs/upgrade

Reply:
Right, mentioning r28 is a little awkward as it's not a stable release. I've tweaked the wording to use r27 as the pivot instead and fixed the hyperlinks (this change only affected 3 pages but made me feel very sorry for web developers that work on refactorings for big websites :rofl: )

Reply:
I'm happy with the documentation updates. I don't have a privilege to mark the conversation as resolved since I'm not a collaborator. But feel free to close this.